web/src/main/java/org/springframework/security/web/server/authentication/SwitchUserWebFilter.java

```diff
@@ -0,0 +1,366 @@
+		Assert.notNull(userDetailsService, "userDetailsService must be specified");
+		Assert.notNull(successHandler, "successHandler must be specified");
+
+		this.userDetailsService = userDetailsService;
+		this.successHandler = successHandler;
+		this.failureHandler = failureHandler;
+
+		this.securityContextRepository = new WebSessionServerSecurityContextRepository();
+		this.userDetailsChecker = new AccountStatusUserDetailsChecker();
+	}
+
+	/**
+	 * Creates a filter for the user context switching
+	 *
+	 * @param userDetailsService The <tt>UserDetailService</tt> which will be used to load
+	 *                           information for the user that is being switched to.
+	 * @param successTargetUrl   Sets the URL to go to after a successful switch / exit user request
+	 * @param failureTargetUrl   The URL to which a user should be redirected if the switch fails
+	 */
+	public SwitchUserWebFilter(ReactiveUserDetailsService userDetailsService,
+			String successTargetUrl, @Nullable String failureTargetUrl) {
+		Assert.notNull(userDetailsService, "userDetailsService must be specified");
+		Assert.notNull(successTargetUrl, "successTargetUrl must be specified");
+
+		this.userDetailsService = userDetailsService;
+		this.successHandler = new RedirectServerAuthenticationSuccessHandler(successTargetUrl);
+		this.failureHandler = Optional.ofNullable(failureTargetUrl).
+				map(RedirectServerAuthenticationFailureHandler::new).orElse(null);
+
+		this.securityContextRepository = new WebSessionServerSecurityContextRepository();
+		this.userDetailsChecker = new AccountStatusUserDetailsChecker();
+	}
+
+	@Override
+	public Mono<Void> filter(ServerWebExchange exchange, WebFilterChain chain) {
+		final WebFilterExchange webFilterExchange = new WebFilterExchange(exchange, chain);
+
+		return switchUser(webFilterExchange)
+				.switchIfEmpty(Mono.defer(() -> exitSwitchUser(exchange)))
+				.switchIfEmpty(Mono.defer(() -> chain.filter(exchange).then(Mono.empty())))
+				.flatMap(authentication -> onAuthenticationSuccess(authentication, webFilterExchange))
+				.onErrorResume(SwitchUserAuthenticationException.class, exception -> Mono.empty());
+	}
+
+	protected Mono<Authentication> switchUser(WebFilterExchange webFilterExchange) {
+		return switchUserMatcher.matches(webFilterExchange.getExchange())
+				.filter(ServerWebExchangeMatcher.MatchResult::isMatch)
+				.flatMap(matchResult -> ReactiveSecurityContextHolder.getContext())
+				.map(SecurityContext::getAuthentication)
+				.flatMap(currentAuthentication -> {
+					final String username = getUsername(webFilterExchange.getExchange());
+					return attemptSwitchUser(currentAuthentication, username);
+				})
+				.onErrorResume(AuthenticationException.class, e ->
+						onAuthenticationFailure(e, webFilterExchange)
+								.then(Mono.error(new SwitchUserAuthenticationException(e)))
+				);
+	}
+
+	protected Mono<Authentication> exitSwitchUser(ServerWebExchange exchange) {
```

Comment:
It seems odd that the parameters for `switchUser` and `exitSwitchUser` are different.
I think it makes more sense if they both take in a `WebFilterExchange`.

Reply:
done